flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java

```diff
@@ -0,0 +1,263 @@
+
+                    SequenceNumber upTo = stateChangelogWriter.lastAppendedSequenceNumber().next();
+
+                    if (upTo.equals(materializedState.lastMaterializedTo())) {
+                        return;
+                    }
+
+                    RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture =
+                            keyedStateBackend.snapshot(
+                                    materializedId++, // TO DISCUSS: seems this does not matter,
+                                    // based on
+                                    System.currentTimeMillis(),
+                                    streamFactory,
+                                    CHECKPOINT_OPTIONS);
+
+                    // TODO: add metadata to log FLINK-23170
+                    asyncOperationsThreadPool.execute(
+                            () -> asyncMaterialization(materializedRunnableFuture, upTo));
+                },
+                "materialization");
+    }
+
+    private void asyncMaterialization(
+            RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture,
+            SequenceNumber upTo) {
+
+        FileSystemSafetyNet.initializeSafetyNetForThread();
+        try {
+            FutureUtils.runIfNotDoneAndGet(materializedRunnableFuture);
+            mailboxExecutor.execute(
+                    () -> {
+                        materializedState =
+                                new MaterializedState(
+                                        getMaterializedResult(materializedRunnableFuture.get()),
+                                        new ArrayList<>(),
+                                        upTo);
+
+                        checkState(
+                                materializationOnGoing.compareAndSet(true, false),
+                                "expect to finish materialization successfully, "
+                                        + "flag materializationOnGoing should be true before finishing.");
+                        retries.set(0);
+                    },
+                    "update materializedSnapshot up to changelog sequence number: {}",
+                    upTo);
+        } catch (Exception e) {
+            int retryTime = retries.incrementAndGet();
+
+            LOG.info(
+                    "Asynchronous part of materialization could not be completed for the {} time.",
+                    retryTime,
+                    e);
+
+            // ToDO: Double check
+            //  how the async phase is related to materialized state tracking
+            //  refer to AsyncCheckpointRunnable#run#catch
+            //  cleanup logic in StateUtil#discardStateFuture
+            handleExecutionException(materializedRunnableFuture);
+
+            if (retryTime == allowedNumberOfFailures) {
+                asyncExceptionHandler.handleAsyncException(
+                        "Fail to complete the asynchronous part of materialization", e);
+            }
```

Comment:
1. I think we should either reset failure counter here or check it before each materialization.. Otherwise, we depend on asyncExceptionHandler for consistent behaviour.
2. Replace `==` with `>=`?

Reply:
>=; done!

what do you mean by `asyncExceptionHandler` consistent behavior? why should we reset to 0 if we want to fail reaching max retry number?

Why do I have to check the max number before materialization (sync phase I guess you mean) before it fails? Doesn't it can bump up only if a failure happens?

Reply:
Maybe let's discuss first what should happen with subsequent materialization once `numberOfConsecutiveFailures` reaches the limit?

If we want them to run, should they have the same (configured) number of attempts? Then we should reset it here?
If we don't, i.e. we prevent any further materializations, then we should check this counter before starting materialization?

Currently, the latter option I think is enforced implicitly by `asyncExceptionHandler` - which fails the job.

Reply:
The contract is:,

If reaching the limit (unrecoverable), we should fail the task.

`asyncExceptionHandler` is a way to fail the task.

resetting te counter is misleading and confusing to this contract.